### DISCUSSION TITLE
Depends doesn't work with contextvars

### DISCUSSION THREAD

**Discussion by norbert-sebok** [fastapi/fastapi#10608] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import contextvars

import uvicorn
from fastapi import Depends, FastAPI

app = FastAPI()
context_key = contextvars.ContextVar("key", default="default")


def save_key(key: str):
    context_key.set(key)


@app.post("/without_depends")
async def without_depends(key: str):
    context_key.set(key)
    return context_key.get()


@app.post("/with_depends", dependencies=[Depends(save_key)])
async def with_depends():
    return context_key.get()


uvicorn.run(app)
```

### Description

An endpoint with a query parameter echos the value perfectly:

```
>>> requests.post("http://127.0.0.1:8000/without_depends?key=abcd").content
b'"abcd"'
```

But the other endpoint which uses Depends returns the default value of the ContextVar:

```
>>> requests.post("http://127.0.0.1:8000/with_depends?key=abcd").content
b'"default"'
```

When I print the context_key object then it has the same object id everywhere.

Am I missing something or the Depends machinery uses a different async context?

### Operating System

Linux

### FastAPI Version

0.104.1

### Pydantic Version

1.10.13

### Python Version

Python 3.8.18

**Comment by norbert-sebok** [comment, marked as answer]
It turned out that the Depends functions have to be async. If I change it to:

```
async def save_key(key: str):
    ...
```
Then it works perfectly.

**Comment by tomkcook** [comment]
Although you found the answer you were looking for, describing this as "perfect" is still a bit of a stretch.  For instance, if you change your `save_key` dependency to this:

```
async def save_key(key: str):
    try:
        token = context_key.set(key)
        yield key
    finally:
        context_key.reset(token)
```
Then you get an error along these lines:
```
  File "/home/tkcook/Scratch/fastapi/venv/lib/python3.10/site-packages/fastapi/concurrency.py", line 37, in contextmanager_in_threadpool
    await anyio.to_thread.run_sync(
  File "/home/tkcook/Scratch/fastapi/venv/lib/python3.10/site-packages/anyio/to_thread.py", line 56, in run_sync
    return await get_async_backend().run_sync_in_worker_thread(
  File "/home/tkcook/Scratch/fastapi/venv/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 2441, in run_sync_in_worker_thread
    return await future
  File "/home/tkcook/Scratch/fastapi/venv/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 943, in run
    result = context.run(func, *args)
  File "/home/tkcook/.pyenv/versions/3.10.14/lib/python3.10/contextlib.py", line 142, in __exit__
    next(self.gen)
  File "/home/tkcook/Scratch/fastapi/app.py", line 15, in save_key
    context_key.reset(token)
ValueError: <Token var=<ContextVar name='key' default='default' at 0x7b05a7144cc0> at 0x7b05a704a8c0> was created in a different Context
```

I'm struggling rather to see how this can happen (though I haven't looked at the FastAPI source yet to try, either).

**Reply by norbertsebok** [reply]
Could you paste the whole file? If I change only the `save_key` function then it doesn't raise that exception. I'm wondering what could create that other Context.

Or maybe the version is the different. Which fastapi version do you use?
